src/smartpeak/source/ui/GraphicDataVizWidget.cpp

```diff
@@ -0,0 +1,256 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey, Ahmed Khalil, Bertrand Boudaud $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
+// --------------------------------------------------------------------------
+
+#include <SmartPeak/ui/GraphicDataVizWidget.h>
+#include <imgui.h>
+#include <imgui_internal.h>
+#include <implot.h>
+
+namespace SmartPeak
+{
+
+  void GraphicDataVizWidget::draw()
+  {
+    updateData();
+    drawSliders();
+    drawGraph();
+  }
+
+  void GraphicDataVizWidget::drawSliders()
+  {
+    // Widget's controls - that ImGui does not support natively
+    const ImGuiSliderFlags slider_flags = ImGuiSliderFlags_AlwaysClamp;
+    float controls_pos_start_y = ImGui::GetCursorPosY();
+    ImGui::SliderFloat((std::string("min ") + chrom_.x_axis_title_).c_str(), &current_range_.first, slider_min_max_.first, current_range_.second, "%.4f", slider_flags);
+    ImGui::SameLine();
+    ImGui::Checkbox("Compact View", &compact_view_);
+    ImGui::SliderFloat((std::string("max ") + chrom_.x_axis_title_).c_str(), &current_range_.second, current_range_.first, slider_min_max_.second, "%.4f", slider_flags);
+    ImGui::SameLine();
+    ImGui::Checkbox("Legend", &show_legend_);
+    if (chrom_.z_axis_title_)
+    {
+      std::ostringstream os;
+      os << chrom_.z_data_area_[current_z_];
+      ImGui::SliderInt(chrom_.z_axis_title_->c_str(), &current_z_, 0, chrom_.z_data_area_.size() - 1, os.str().c_str(), ImGuiSliderFlags_NoInput);
+    }
+    float controls_pos_end_y = ImGui::GetCursorPosY();
+    sliders_height_ = (controls_pos_end_y - controls_pos_start_y);
+  }
+
+  std::tuple<float, float, float, float> GraphicDataVizWidget::plotLimits() const
+  {
+    // compute plot limit by adding padding
+    float border_padding_x = (chrom_.x_max_ - chrom_.x_min_) * 0.01f;
+    float border_padding_y = (chrom_.y_max_ - chrom_.y_min_) * 0.01f;
+    auto plot_min_x = current_range_.first - border_padding_x;
+    auto plot_max_x = current_range_.second + border_padding_x;
+    auto plot_min_y = chrom_.y_min_ - border_padding_y;
+    auto plot_max_y = chrom_.y_max_ * 1.1f;
+    return std::make_tuple(plot_min_x, plot_max_x, plot_min_y, plot_max_y);
+  }
+
+  void GraphicDataVizWidget::drawGraph()
+  {
+    // Main graphic
+    if (chrom_.points_overflow)
+    {
+      ImGui::Text("Unable to draw: too much points. Please reduce scope or unselect data.");
+    }
+    else if (chrom_.x_data_area_.empty() && chrom_.y_data_area_.empty())
+    {
+      ImGui::Text("No data to display select data to render or adjust ranges.");
+    }
+    else
+    {
+      auto [plot_min_x, plot_max_x, plot_min_y, plot_max_y] = plotLimits();
```

Comment:
Nice one liner :)

Reply:
yes indeed :) very nice to use modern c++.